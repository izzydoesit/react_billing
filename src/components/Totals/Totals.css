div.totals-wrapper {
  margin: 1%;
  padding: 1%;
}

p {
  padding: 2%;
  border: 1px solid #12c8c1;
  background-color: #e5fdf9;
  width: 20%;
  color: #949598
}

.danger {
  border: 2px solid #f86028;
}

.success {
  border: 2px solid #08C6BE;
}

.pending {
  border: 2px solid #FEC87D;
}