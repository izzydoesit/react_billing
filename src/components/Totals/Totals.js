// @flow 

import React, { Component } from 'react';
import {
  isToday,
  isPastDate,
  isFutureDate,
  formatCentsToDollars } from 'clark-utils';
import type { BillsType } from '../../types'
import './Totals.css';

type Props = {
  bills: BillsType,
};

export default class Totals extends Component<Props> {
  static defaultProps = {
    bills: {}
  }

  calculateTotals = (bills: BillsType): string => {
    let total = 0, paid = 0, overdue = 0, outstanding = 0;

    for (let key in bills) {
      
      const bill = bills[key];
      const billDueDate = bill['dueDate'];

      if (bill && (bill['status'] === 'pending' && isPastDate(billDueDate))) {
        overdue += bill['amountInCents'];
      } else if (bill['status'] === 'pending' && (isToday(billDueDate) || isFutureDate(billDueDate))) {
        outstanding += bill['amountInCents'];
      } else if (bill['status'] === 'paid') {
        paid += bill['amountInCents'];
      } 
    };

    total = (paid + overdue + outstanding);

    return { total, paid, overdue, outstanding};
  }

  render() {
      const bills = this.props.bills;
      const { total,
              paid, 
              overdue, 
              outstanding} = this.calculateTotals(bills)


    return (
      <div className="totals-wrapper">
        <p className="all-total">Total: <span className="number-total">{formatCentsToDollars(total)}</span></p>
        <p className="paid-total success">Paid: <span className="number-total">{formatCentsToDollars(paid)}</span></p>
        <p className="overdue-total danger">Overdue: <span className="number-total">{formatCentsToDollars(overdue)}</span></p>
        <p className="outstanding-total pending">Outstanding: <span className="number-total">{formatCentsToDollars(outstanding)}</span></p>
      </div>
    )
  }
}
