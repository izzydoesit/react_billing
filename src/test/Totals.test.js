// @flow

import React from 'react';
import Totals from '../components/Totals/Totals';
import billData from '../data';
import { mount, configure } from 'enzyme';
import { mountToJson } from 'enzyme-to-json';
import Adapter from 'enzyme-adapter-react-16';
configure({ adapter: new Adapter() });

describe('<Totals />', () => {
  let props, list;
  let mountedTotals;
  const totals = () => {
    if (!mountedTotals) {
      mountedTotals = mount(<Totals {...props} />);
    }
    return mountedTotals;
  };

  beforeEach(() => {
    props = {
      bills: {}
    }
    mountedTotals = undefined;
  });

  it('renders correctly', () => {
    expect(mountToJson(totals())).toMatchSnapshot();
  });

  it('renders a <div> element', () => {
    expect(totals().find('div.totals-wrapper').length).toEqual(1);
  });

  it('renders a Total <p> element', () => {
    expect(totals().find('p.all-total').length).toEqual(1);
    expect(totals().find('p.all-total').text()).toContain('Total:');
  });

  it('renders a Paid <p> element', () => {
    expect(totals().find('p.paid-total').length).toEqual(1);
    expect(totals().find('p.paid-total').text()).toContain('Paid:');
  });

  it('renders an Overdue <p> element', () => {
    expect(totals().find('p.overdue-total').length).toEqual(1);
    expect(totals().find('p.overdue-total').text()).toContain('Overdue:');
  });

  it('renders an Outstanding <p> element', () => {
    expect(totals().find('p.outstanding-total').length).toEqual(1);
    expect(totals().find('p.outstanding-total').text()).toContain('Outstanding:');
  });

  describe('when bills is defined', () => {
    beforeEach(() => {
      props.bills = billData;
    });

    it('renders a total of all bill amounts', () => {
      expect(totals().find('p.all-total > span.number-total').text()).toEqual('$16,989.55');
    });

    it('renders a total of all paid bill amounts', () => {
      expect(totals().find('p.paid-total > span.number-total').text()).toEqual('$1,779.23');
    });

    it('renders a total of all outstading bill amounts', () => {
      expect(totals().find('p.overdue-total > span.number-total').text()).toEqual('$6,305.42');
    });

    it('renders a total of all overdue bill amounts', () => {
      expect(totals().find('p.outstanding-total > span.number-total').text()).toEqual('$8,904.90');
    });
  });
});
