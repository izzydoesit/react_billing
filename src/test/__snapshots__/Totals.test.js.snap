// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Totals /> renders correctly 1`] = `
<Totals
  bills={Object {}}
>
  <div
    className="totals-wrapper"
  >
    <p
      className="all-total"
    >
      Total: 
      <span
        className="number-total"
      >
        0
      </span>
    </p>
    <p
      className="paid-total success"
    >
      Paid: 
      <span
        className="number-total"
      >
        0
      </span>
    </p>
    <p
      className="overdue-total danger"
    >
      Overdue: 
      <span
        className="number-total"
      >
        0
      </span>
    </p>
    <p
      className="outstanding-total pending"
    >
      Outstanding: 
      <span
        className="number-total"
      >
        0
      </span>
    </p>
  </div>
</Totals>
`;
